### ISSUE TITLE
nvmf_tgt/rdma: Lost NVMe commands with SRQ after nvmf_tgt restarts

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#2876]
# Sighting report

I've run into a strange behavior testing the kernel initiator and nvmf_tgt where we seem to be losing NVMe commands on the target side after it was restarted.

The scenario is as follows:
- Kernel initiator is connected to 3 targets, all targets and initiator are on different physical systems.
- Systems are connected via a switch, set up for lossless RDMA traffic
- Each of the targets have 24 subsystems, not all subsystems have namespaces (32 namespaces total)
- The nvme connect is pretty basic: nvme connect -k 20 -c 5 -T 184 -a $IOM -t rdma -s 4420 -i 1 -n $nqn
- No IO load is running

Connecting to all 3x24 subsystems works as expected. IO is fine as well. However, when I restart the **all** 3 targets, then sometimes we are losing NVMe commands.

When the target is restarted, the kernel initiator performs its reconnect work. This all works as expected. I can see the "Successfully reconnected (3 attempts)" message which indicates that the initiator was able to init the ctrlr and create the IO qpair. However, occasionally, the initiator is unable to receive a keep-alive completion for the device (nvmeXYZ) that was just reconnected successfully.

Note, this issue doesn't happen for all devices... out of the 144 devices, I've only seen 1-4 devices showing this behavior. It's also different devices/subsystems every time. But it's always reproducible.

I added some tracing to kernel and SPDK. I see that the WR is created for the keep alive and it is submitted via IB verbs. On the SPDK side, I instrumented all incoming RDMA capsules and I do **NOT** see the keep alive capsule coming in when the kernel shows the timeout for the keep alive for the corresponding device/qpair.
It's also not an issue related to keep alive ... I saw this behavior with IOs as well and an artificially injected identify a couple seconds after the reconnect succeeded.

Now, when I disable shared receive queue (no_srq true), this issue goes away. So it seems to be related to the SRQ handling within SPDK or HBA/driver.

## Expected Behavior

SRQ should behave the same way as no SRQ.

## Steps to Reproduce

It's probably best to do this in a physical setup but it's complex since it requires may switch.
In my setup, I connect the kernel initiator with the "nvme connect" command to 3x24 subsystems, one by one.
Then I restart all the targets. The kernel initiator will try to reconnect and this is where sometimes NVMe commands are lost for certain subsystems (as indicated by the keep alive timeouts).

## Context (Environment including OS version, SPDK version, etc.)

SPDK 22.01
Debian 11.3 (initiator), Debian 11.2 (target)
Mellanox CX6-DX with latest firmware (22.35.2000)

**Comment by mhae** [comment]
FWIW, here are the kernel trace messages with some instrumentation I added.

```
Jan 18 23:04:52 localhost vmunix: [  111.976589] nvme nvme95: Reconnecting in 5 seconds...
// Next reconnect attempt was successful. The admin queue and an IO queue was created.
Jan 18 23:04:57 localhost vmunix: [  117.202096] nvme nvme95: ???: QID 0 comp vector 0
Jan 18 23:04:57 localhost vmunix: [  117.240596] nvme nvme95: creating 1 I/O queues.
Jan 18 23:04:57 localhost vmunix: [  117.896732] nvme nvme95: ???: QID 1 comp vector 20
Jan 18 23:04:58 localhost vmunix: [  118.079547] nvme nvme95: mapped 1/0/0 default/read/poll queues.
// The above message shows that connection to the subsystem was successful.

Jan 18 23:04:58 localhost vmunix: [  118.080212] nvme nvme95: Successfully reconnected (3 attempts)

// At this point a keep alive was posted to the submission queue
Jan 18 23:05:59 localhost vmunix: [  179.265976] nvme nvme95: ??? sending keep-alive

// 18s later we're receiving a disconnect initiated by the target due to keep alive timeout.
Jan 18 23:06:17 localhost vmunix: [  197.096043] nvme nvme95: ??? disconnect received - connection closed [ev->event=10] - starting error recovery
Jan 18 23:06:17 localhost vmunix: [  197.096048] nvme nvme95: starting error recovery
Jan 18 23:06:17 localhost vmunix: [  197.096389] nvme nvme95: ??? disconnect received - connection closed [ev->event=10] - starting error recovery
Jan 18 23:06:17 localhost vmunix: [  197.250742] nvme nvme95: failed nvme_keep_alive_end_io error=10 [kato=60] [ctrl->state=2]
// Reconnect begins again ...
Jan 18 23:06:17 localhost vmunix: [  197.272424] nvme nvme95: Reconnecting in 5 seconds...
```

**Comment by mhae** [comment]
I wasn't able to reproduce this issue outside of that one setup where it is failing. I tried the following:
- Changing nvmf cores to 1 seems to make it worse (more commands lost)
- I wasn't able to repro the issue when I only restart one target (instead of two)
- I set up another environment where I have one physical initiator and 2 physical targets. I ran 2 target apps on each of the physical systems. I used 48 subsystems/0drives each target simulated with AIO. I wasn't able to repro the issue in this env either. One of the differences (besides the initiator HW) is that this env I was using CX5s and in the other CX6DX.

**Comment by mhae** [comment]
Today I was able to reproduce the problem in two of my environments with just one target. In one of the setups I just had to add an additional 24 subsystems (witohout namespaces) for a total of 48. In the other setup, I had to create 250 (!) subsystems (most of them empty) to reproduce the issue.

**Comment by AlekseyMarchuk** [comment]
Hi @mhae,

Could you try to collect tcpdumps on the target side to understand if KATO command was transferred to the target host?

**Comment by mhae** [comment]
@AlekseyMarchuk , I can give it a try. Last time I tried to mirror ports and use tcpdump to capture RDMA traffic didn't work.
I can see if tcpdump works on the target.

As mentioned, I'm pretty sure it reaches the qpair since I traced it down the where it is placed in the queue on the initiator side.

**Comment by mhae** [comment]
@AlekseyMarchuk , I've spent all morning to see if we can get tcpdump to work ... it seems we can't use it due to the offload and we'd need to set up port mirroring on the switch. Last time we tried this, we couldn't get this to work either,
I also played with ibdump but I think the one that comes with MLNX_OFED doesn't work with CX5/6 cards.

As I mentioned above, I'm sure the KA cmd wasn't picked up from the WR queue on the target. I did dump all KA commands and I didn't see the ones from the block devices that were failing.

The question is really why SRQ in combination with restarts exhibits this behavior. Do you know if the CM is actively monitoring the connection and has the ability to tear down the connection?

**Comment by AlekseyMarchuk** [comment]
CM doesn't monitor physical connection, that is done on transport level. If there are problems with physical link, requests will be completed (with ibv_poll_cq) with IBV_WC_RETRY_EXC_ERR status.
RDMACM can detect if remote qp was disconnected using rdma_disconnect
If have several questions/suggestions
1. Kernel target also uses SRQ, I don't know what is default option, but it can be enabled via nvmet_rdma parameters. Could you try to reproduce this issue with kernel target? 
2. Does this issue occur on different target hosts? I believe you would notice if happens only on 1 host, that might mean a problem in specific NIC, just wanted to clarify
3. If a capsule with KA was lost on a switch or on receiver's NIC then initiator NIC would retransmit the packet. If retransmit count exceeded, we would see an error on initiator side. Since there is no IO, I believe this case is hardly ever possible, but I would suggest to verify that. There are several parameters in rdma_conn_param structure: retry_count - total number of retransmissions if ack is not received; and rnr_retry_count - the number of retransmissions when one side issues a send command but there was no buffer in the receiver's receive queue.  They are hardcoded in kernel initiator: https://github.com/torvalds/linux/blob/22b8077d0fcec86c6ed0e0fce9f7e7e5a4c2d56a/drivers/nvme/host/rdma.c#L1874 Could you try to set them to 1? 
4. Also there is an extended option to set transport ack time for RDMA connection - that is not used by kernel initiator but it would be great if you could add it. Please refer to https://github.com/spdk/spdk/commit/94966468ae249c769795a8e40ffd94d2bc3bda62 as an example. Here are rough measurements of different values of ack_timeout https://github.com/spdk/spdk/issues/1124#issuecomment-568695203 By setting small values of ack_timeout and retry counts we can receive a RETRY_EXC error earlier which might point on problems in network or in NIC itself.
5. If possible, try to exclude a switch from your setup

**Comment by mhae** [comment]
@AlekseyMarchuk , thanks for your pointers and sorry for the late reply.

Re 1: Unfortunately, I don't have the nvme target compiled into the kernel that we're using on the target. It would be some effort to get this going.
Re 2: It's happening on any target and I was able to reproduce it in 2 environments.
Re 3: I made this change but nothing really changed. Same issue and dmesg output.
Re 4: Same as 3. No change. I made the change in cma_connect_ib:
![image](https://user-images.githubusercontent.com/6045616/216469319-9d98861b-ddae-4960-9b05-e60795d1ba6b.png)
Re 5: I could do that but I would need to make sure I can repro it with one target. However, given that I don't see any issue with high IO loads through the switch during normal operations, I would think that the switch isn't the culprit here.

Now, I made a change to the kernel where I throttle the number of concurrent re-connects (nvme_rdma_reconnect_ctrl_work). When I limit the # of outstanding re-connects to 24, then I don't see the issue anymore. I had to introduce the throttling for the non-SRQ case when we have a large number of subsystems (600 range) that needed a re-connect (basically in the non-SRQ case, the kernel was never able to re-connect all subsystems and spun forever).

It seems to me that if too many re-connects are happening then we are running into some kind of bottleneck. I can't say for sure whether using SRQ with small number of subsystems or non-SRQ with large number of subsystems have the same root cause but the workaround fixes both issues.

It really bugs me that I can't pin down the root cause. I'm somewhat worried that these issues show up in another setup that is slightly different.
